pkg/cloudprovider/provider/linode/provider.go

```diff
@@ -0,0 +1,447 @@
+			glog.V(6).Infof("linode (id='%d') got fully created", linode.ID)
+			return true, nil
+		}
+		glog.V(6).Infof("waiting until linode (id='%d') got fully created...", linode.ID)
+		return false, nil
+	})
+
+	return &linodeInstance{linode: linode}, err
+}
+
+func (p *provider) Cleanup(machine *v1alpha1.Machine, _ *cloud.MachineCreateDeleteData) (bool, error) {
+	instance, err := p.Get(machine)
+	if err != nil {
+		if err == cloudprovidererrors.ErrInstanceNotFound {
+			return true, nil
+		}
+		return false, err
+	}
+
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return false, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+	ctx := context.TODO()
+	client := getClient(c.Token)
+
+	linodeID, err := strconv.Atoi(instance.ID())
+	if err != nil {
+		return false, fmt.Errorf("failed to convert instance id %s to int: %v", instance.ID(), err)
+	}
+
+	err = client.DeleteInstance(ctx, linodeID)
+	if err != nil {
+		return false, doStatusAndErrToTerminalError(err)
+	}
+
+	return false, nil
+}
+
+func (p *provider) Get(machine *v1alpha1.Machine) (instance.Instance, error) {
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	ctx := context.TODO()
+	client := getClient(c.Token)
+
+	filter, _ := json.Marshal(map[string]interface{}{
+		"label": machine.Spec.Name,
+	})
+
+	listOptions := linodego.NewListOptions(0, string(filter))
+	linodes, err := client.ListInstances(ctx, listOptions)
+
+	if err != nil {
+		return nil, doStatusAndErrToTerminalError(err)
+	}
+
+	for i, linode := range linodes {
+		if linode.Label == machine.Spec.Name && sets.NewString(linode.Tags...).Has(string(machine.UID)) {
+			return &linodeInstance{linode: &linodes[i]}, nil
+		}
+	}
+
+	return nil, cloudprovidererrors.ErrInstanceNotFound
+}
+
+func (p *provider) MigrateUID(machine *v1alpha1.Machine, new types.UID) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to decode providerconfig: %v", err)
+	}
+	client := getClient(c.Token)
+	linodes, err := client.ListInstances(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("failed to list linodes: %v", err)
+	}
+
+	for _, linode := range linodes {
+		if linode.Label == machine.Spec.Name && sets.NewString(linode.Tags...).Has(string(machine.UID)) {
+			updateOpts := linode.GetUpdateOptions()
+
+			tags := []string{string(new)}
+			if updateOpts.Tags != nil {
+				oldUID := string(machine.UID)
+				for _, existingTag := range *updateOpts.Tags {
+					if existingTag != oldUID {
+						tags = append(tags, existingTag)
```

Comment:
Can we remove the old tag here?

Reply:
Unless I missed something, that's what I'm doing.

I start with `tags := []string{string(new)}` and just add back all of the other tags, except for the `oldUID` tag.

Reply:
Oh you are right, my bad

Reply:
UpdateInstance returns the full Linode instance object, I'll change this to not throw it away since it is used below.

I'll also need to change those status strings to match the Linode ones (there are consts to be used there).

Reply:
I misread the context from my phone.. I updated the status strings.  There is no reason to keep the linode instance object returned from UpdateInstance, here.